### ISSUE TITLE
assertWarns has some restrictions

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2703]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2703|
|Type|enhancement|
|Created|2007-06-11 09:50:25Z|

First uses of assertWarns have shown some limits:
 * it can't be used to check warnings emitted by class instantiation or implicit !__call!__ method
 * it doesn't work with warnings filtered by default like PendingDeprecationWarning

<details><summary>Searchable metadata</summary>

```
trac-id__2703 2703
type__enhancement enhancement
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1181555425000000 1181555425000000
changetime__1182674159000000 1182674159000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve removed owner|
|-|-|

Ready to review in assertwarns-features-2703.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

What exactly do you mean by "it can't be used with class"?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

I've modified the description to reflect what I meant.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Arg, I hate the stupid warnings module.

Noble effort, but the only actually correct way to get !__warningsregistry!__ is to go through the same dance as warnings.warn.  Even if we duplicate it, there's no guarantee Python devs won't change it randomly for the next release.

I'm trying to remember why we don't do something like patch ```warnings.warn_explicit``` instead of trying to re-assemble the events after the fact.  I think the reason we didn't patch warnings.warn was to support ```from warnings import warn```, but I can't think of a similar reason not to patch ```warn_explicit```.

(This is not a review, just some thoughts.)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

We dont't patch warn_explicit because _we_ didn't implemented like that at the beginning :). I guess that could work though.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

I've tried it, and it's too good to be true :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Hmm, for me, some tests in twisted.trial.test.test_tests fail w/ this branch merged into trunk, mostly similar to this:

```
===============================================================================
[FAIL]: twisted.trial.test.test_tests.SuppressionTest.testSuppressModule

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/test/test_tests.py", line 507, in testSuppressModule
    self.assertSubstring(suppression.METHOD_WARNING_MSG, self.getIO())
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/unittest.py", line 307, in failUnlessSubstring
    return self.failUnlessIn(substring, astring, msg)
twisted.trial.unittest.FailTest: 'method warning message' not in ''
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Damn, sorry, that was something stupid.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

No problem.  Looks great, please merge :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [20610]) Merge assertwarns-features-2703

Author: therve
Reviewer: exarkun
Fixes [#2703](https://github.com/twisted/twisted/issues/2703)

Fix implementation of assertWarns to work with all callables and with all
type of warnings.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
